src/Gui/ShortcutManager.cpp

```diff
@@ -0,0 +1,455 @@
+            if (!accel)
+                accel = "";
+        }
+    }
+    QString shortcut;
+    if (cmdName)
+        shortcut = QString::fromLatin1(hShortcuts->GetASCII(cmdName, accel).c_str());
+    else
+        shortcut = QString::fromLatin1(accel);
+    return QKeySequence(shortcut).toString(QKeySequence::NativeText);
+}
+
+void ShortcutManager::setShortcut(const char *cmdName, const char *accel)
+{
+    if (cmdName && cmdName[0]) {
+        setTopPriority(cmdName);
+        if (!accel)
+            accel = "";
+        if (auto cmd = Application::Instance->commandManager().getCommandByName(cmdName)) {
+            auto defaultAccel = cmd->getAccel();
+            if (!defaultAccel)
+               defaultAccel = "";
+           if (QKeySequence(QString::fromLatin1(accel)) == QKeySequence(QString::fromLatin1(defaultAccel))) {
+                hShortcuts->RemoveASCII(cmdName);
+                return;
+           }
+        }
+        hShortcuts->SetASCII(cmdName, accel);
+    }
+}
+
+bool ShortcutManager::checkShortcut(QObject *o, const QKeySequence &key)
+{
+    auto focus = QApplication::focusWidget();
+    if (!focus)
+        return false;
+    auto action = qobject_cast<QAction*>(o);
+    if (!action)
+        return false;
+
+    const auto &index = actionMap.get<1>();
+    auto iter = index.lower_bound(ActionKey(key));
+    if (iter == index.end())
+        return false;
+
+    auto it = iter;
+    // skip to the first not exact matched key
+    for (; it != index.end() && key == it->key.shortcut; ++it);
+
+    // check for potential partial match, i.e. longer key sequences
+    bool flush = true;
+    for (; it != index.end(); ++it) {
+        if (key.matches(it->key.shortcut) == QKeySequence::NoMatch)
+            break;
+        if (it->action && it->action->isEnabled()) {
+            flush = false;
+            break;
+        }
+    }
+
+    int count = 0;
+    for (it = iter; it != index.end() && key == it->key.shortcut; ++it) {
+        if (it->action && it->action->isEnabled()) {
+            if (!flush) {
+                // temporary disable the action so that we can try potential
+                // match with further keystrokes.
+                it->action->setEnabled(false);
+            }
+            pendingActions.emplace_back(it->action, key.count(), getPriority(it->key.name));
+            ++count;
+        }
+    }
+    if (!count) {
+        // action not found in the map, shouldn't happen!
+        pendingActions.emplace_back(action, key.count(), 0);
+    }
+    if (flush) {
+        // We'll flush now because there is no potential match with further
+        // keystrokes, so no need to wait for timer.
+        lastFocus = nullptr;
+        onTimer();
+        return true;
+    }
+
+    lastFocus = focus;
+    pendingSequence = key;
+
+    // Qt's shortcut state machine favors shortest match (which is ridiculous,
+    // unless I'm mistaken?). We'll do longest match. We've disabled all
```

Comment:
I don't know if you're mistaken, but I agree with you that longest match should be the correct behavior.